### PR TITLE
Bug #15092: Fix cache leak between transactions

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

**Comment by fppt** [comment]
@aelred keen eyes Mr.Chapman, keen and supremely picky. This one has the potential to hurt us all.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ConceptCache.java [hunk 1]
```diff
@@ -39,8 +39,14 @@
  *
  */
 class ConceptCache<V> {
+    //If no cache can produce the data then the database is read
     private final Supplier<V> databaseReader;
-    private Optional<V> cachedValue = Optional.empty();
+
+    //Transaction bound. If this is not set it does not yet exist in the graph
+    private ThreadLocal<V> cachedValue = new ThreadLocal<>();
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ConceptCache.java]: Not sure what this comment means. Isn't it possible for something to exist in the graph, but not be set in the cache?

File: grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java [hunk 2]
```diff
@@ -123,10 +123,16 @@
...
+    /**
+     * Flushes the internal transaction caches so that persisted information can be cached and shared between
+     * concepts
+     */
+    public void flushTxCache(){
+        cachedIsImplicit.flush();
+        cachedIsAbstract.flush();
+        cachedSuperType.flush();
+        cachedDirectSubTypes.flush();
+        cachedShards.flush();
```
Comment (aelred) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java]: `cachedDirectPlays.flush()` needs to be here.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/ConceptCache.java (hunk 1 above)
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/ConceptCache.java]: yes you right. The comment is misleading.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java (hunk 2 above)
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/TypeImpl.java]: Good eyes.

**Pull request merged**